### PR TITLE
feat(FN-3519): take into account effective date of amendments

### PR DESCRIPTION
**Pull request by RLavender98**
## Introduction :pencil2:
Currently we use latest amendment in utilisation reporting calculations and displaying certain fields, however we actually need to use whatever value is in effect at the report period end.

## Resolution :heavy_check_mark:
- Update amendment helpers to take in an 'effective at' date so that we fetch amendments that are effective for the given date rather than the latest.
- In fixed fee calculation use the cover end date effective at the end of the report period
- In utilisation tab facility value display the value effective at the end of the report period

## Miscellaneous 

This in draft state to get initial review whilst finishing off api and e2e test changes.

### CODE REVIEW

File: dtfs-central-api/api-tests/v1/utilisation-reports/post-keying-data.api-test.ts [hunk 1]
```diff
@@ -316,6 +319,87 @@ describe(`POST ${BASE_URL}`, () => {
...
+    expect(entities).toHaveLength(1);
+    expect(entities[0].id).toEqual(facilityId);
+    /**
+     * The fixed fee is calculated as follows:
+     * fixed fee = utilisation * bank fee adjustment * interest percentage * days left in cover period / day count basis
+     *           = 1 * 0.9 * 100 / 100 * 730 / 1
+     *           = 0.9 * 730
+     *           = 657
+     */
+    expect(entities[0].fixedFee).toEqual(657);
```
Comment (ChristianMcc) [inline dtfs-central-api/api-tests/v1/utilisation-reports/post-keying-data.api-test.ts:400]: just a note that whoever gets in first out of @Zainzzkk and this will need an update here as cover percentage will now be a thing in the calculation

**Review by ChristianMcc — COMMENTED** [review]
Half way through this and no comments so far, nice work - will finish the review asap this morning

File: dtfs-central-api/src/helpers/amendments/get-effective-facility-value-amendment.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import { filterAndSortCompletedEffectiveAmendments } from './filter-and-sort-completed-effective-amendments';
+
+/**
+ * Get the completed amendment facility value that is effective
+ * on the effective at date if there are any
+ * @param tfmFacility - The tfm facility
+ * @param effectiveAtDate - The date to find the effective amendment for
+ * @returns The latest completed amendment cover end date
+ */
+export const getEffectiveFacilityValueAmendment = (tfmFacility: TfmFacility, effectiveAtDate: Date): number | null => {
```
Comment (ChristianMcc) [inline dtfs-central-api/src/helpers/amendments/get-effective-facility-value-amendment.ts:11]: does this still make sense as getLatestEffectiveFacilityValueAmendment? Ditto for the other function

File: dtfs-central-api/src/services/state-machines/fee-record/event-handlers/helpers/calculate-fixed-fee.ts [hunk 3]
```diff
@@ -28,7 +28,7 @@ const getNumberOfDaysRemainingInCoverPeriod = (reportPeriodEnd: MonthAndYear, co
  * @returns The fixed fee for the current report period
  */
 export const calculateFixedFee = ({ utilisation, reportPeriod, coverEndDate, interestPercentage, dayCountBasis }: CalculateFixedFeeParams): number => {
-  const numberOfDaysRemainingInCoverPeriod = getNumberOfDaysRemainingInCoverPeriod(reportPeriod.start, coverEndDate);
+  const numberOfDaysRemainingInCoverPeriod = getNumberOfDaysRemainingInCoverPeriod(reportPeriod.end, coverEndDate);
```
Comment (ChristianMcc) [inline dtfs-central-api/src/services/state-machines/fee-record/event-handlers/helpers/calculate-fixed-fee.ts:31]: you  saviour

**Review by ChristianMcc — COMMENTED** [review]
Really nice work, left a couple of minor nit-picks!

Is there any e2e test that checks generated keying sheet work and do I even want to ask about us adding an amendment to one of the facilities in that test?

**Comment by ChristianMcc** [comment]
For the E2E test i mentioned I would be expected seeded data, not a full E2E test that makes a facility and amendment (even though one like that would be cool and extremely long aha)

**Comment by RLavender98** [comment]
> Is there any e2e test that checks generated keying sheet work and do I even want to ask about us adding an amendment to one of the facilities in that test?

@ChristianMcc There are end to end tests for checking keying data generation. The set up is very complex already so I don't want to make it even more complicated by combining more edge cases into the existing test. I didn't add a new journey specifically for it because I felt the api test I added added sufficient coverage from an integration perspective but I can add a new journey relatively easily if we want that?

File: dtfs-central-api/src/helpers/amendments/get-effective-facility-value-amendment.ts (hunk 2 above)
Comment (ChristianMcc) [reply dtfs-central-api/src/helpers/amendments/get-effective-facility-value-amendment.ts:16]: ah yeah that does make sense tbf, i was thinking too much about the meaning of effective date rather than just the fact that effective is it's own meaning that works here :D getting in my own head writing this comment

File: dtfs-central-api/src/helpers/amendments/filter-and-sort-completed-effective-amendments.test.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+      const amendmentUpdatedAtBeforeLatestEffectiveDate = {
+        ...aCompletedTfmFacilityAmendment(),
+        effectiveDate: undefined,
+        updatedAt: beforeLatestEffectiveDate.getTime(),
+      };
+
+      const amendmentUpdatedAtAfterLatestEffectiveDate = {
+        ...aCompletedTfmFacilityAmendment(),
+        effectiveDate: undefined,
```
Comment (ttbarnes) [inline dtfs-central-api/src/helpers/amendments/filter-and-sort-completed-effective-amendments.test.ts:66]: Could these be null instead of undefined? Not a blocker

File: dtfs-central-api/src/helpers/amendments/filter-and-sort-completed-effective-amendments.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+export const filterAndSortCompletedEffectiveAmendments = (tfmFacilityAmendments: TfmFacilityAmendment[], latestEffectiveDate: Date): TfmFacilityAmendment[] => {
+  const filteredAmendments = tfmFacilityAmendments.filter(({ status, effectiveDate, updatedAt, amendmentId }) => {
+    if (status !== AMENDMENT_STATUS.COMPLETED) {
+      return false;
+    }
+
+    const dateToUse = effectiveDate ? new Date(effectiveDate) : new Date(updatedAt);
+
+    if (!isValidDate(dateToUse)) {
+      const invalidField = effectiveDate ? 'effective date' : 'updated at';
```
Comment (ttbarnes) [inline dtfs-central-api/src/helpers/amendments/filter-and-sort-completed-effective-amendments.ts:26]: I'd prefer these to be real camelCase fields, what do you think?

File: dtfs-central-api/src/helpers/amendments/get-effective-cover-end-date-amendment.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  const completedEffectiveAmendments = filterAndSortCompletedEffectiveAmendments(amendments, effectiveAtDate);
+
+  const effectiveAmendmentWithCoverEndDate = completedEffectiveAmendments.find(({ coverEndDate }) => coverEndDate);
+
+  const effectiveCoverEndDateAmendment = effectiveAmendmentWithCoverEndDate?.coverEndDate ?? null;
+  if (!effectiveCoverEndDateAmendment) {
+    return null;
+  }
```
Comment (ttbarnes) [inline dtfs-central-api/src/helpers/amendments/get-effective-cover-end-date-amendment.ts:30]: A new line before the if statements would make for easier reading please 🙏

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-to-fee-record-utilisation.test.ts [hunk 7]
```diff
@@ -51,31 +51,40 @@ describe('map-to-fee-record-utilisation', () => {
...
 
-      const tfmFacility = aTfmFacility();
-      tfmFacility.facilitySnapshot.ukefFacilityId = facilityId;
-      tfmFacility.facilitySnapshot.value = value;
+      // Act
+      mapToFeeRecordUtilisation(feeRecord, tfmFacility, aReportPeriod());
+
+      // Assert
+      expect(getValueAmendmentSpy).toHaveBeenCalledTimes(1);
+      expect(getValueAmendmentSpy).toHaveBeenCalledWith(tfmFacility, expect.any(Date));
```
Comment (ttbarnes) [inline dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-to-fee-record-utilisation.test.ts:72]: Are we able to assert a real date value?

File: dtfs-central-api/src/helpers/amendments/filter-and-sort-completed-effective-amendments.test.ts (hunk 4 above)
Comment (RLavender98) [reply dtfs-central-api/src/helpers/amendments/filter-and-sort-completed-effective-amendments.test.ts:66]: No, the type is of the field is of type `number` or `undefined`, here we explicitly set to undefined because we want to ensure it is undefined in the test rather than not specifying it and relying on the `aTfmFacilityAmendment` setting it to undefined and then coupling the success/failure of this test to that function definition

File: dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-to-fee-record-utilisation.test.ts (hunk 7 above)
Comment (RLavender98) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-to-fee-record-utilisation.test.ts:72]: We do that in the test below. My thought was it would be better to assert the different params called specifically in separate tests to split assertions and keep the tests more specific but I can combine into just one test if that would be preferred?

Comment (ttbarnes) [reply dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-to-fee-record-utilisation.test.ts:72]: Thanks @RLavender98 makes sense as it is

File: dtfs-central-api/src/helpers/amendments/filter-and-sort-completed-effective-amendments.ts (hunk 5 above)
Comment (RLavender98) [reply dtfs-central-api/src/helpers/amendments/filter-and-sort-completed-effective-amendments.ts]: Agreed, I think that is better!
